File: numarical analysis/lec2.py
import math

def fx(x, y):
    return x + 1 + pow(y,2) 

def ahmed():
    print ("this is a program of oylar methods ")
    print()
    x0, y0 = map(float, input("fx(x0) = y0, please enter x0 and y0: ").split())
    a, b = map(float, input("Please enter interval: ").split())
    if ( a > b ):
        temp = a 
        a = b 
        b = temp 
    h = float(input("Please enter step: "))
    if ( h <= 0):
        print ("not valid , step should be greater than 0 ")
        h = float(input("Please enter step: "))
    elif (h > b):
        print ("not valid , step should be less than or equal "  , b )
        h = float(input("Please enter step: "))
            


    x = []
    y = []

    k = x0
    while k <= b:
        x.append(k)
        k += h

    y.append(y0)

    for i in range(1, len(x)):
        y.append(y[i-1] + h * fx(x[i-1], y[i-1]))

    ye1 = []
    yi = []

    yi.append(y0)
    
    
    for j in range(len(x)):
        
        if j > 0:
            yi.append(yi[j-1] + (h / 2) * abs(fx(x[j-1], yi[j-1]) + fx(x[j], ye1[j-1])))
        if j != len(x) - 1 :
            ye1.append(  yi[j] + h * fx(x[j], yi[j]))
      
    print()
    print("by using oylar method:")
    print("at x =", x[i], ", y =", y[i])
    print()
    print("by using modified oylar method:")
    print("at x =", x[i], ", y =", yi[i])
    print()
    print("thank you DR MOUNIS")
    print()

if __name__ == '__main__':
    t = 1
    # t = int(input())
    while t > 0:
        ahmed()
        t -= 1

File: numarical analysis/lec2.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long 
#define nl "\n"
#define oo INT_MAX 
#define all(vec) vec.begin() , vec.end()
#define put(y) return void(cout << y << nl )
#define sz(r) int(r.size())
#define pi  3.14159265358979323846
#define fixed(n) fixed << setprecision(n) 
#define fr(v,n) deque < ll > v(n) ; for(auto &i : v) cin >> i 
#define mod (ll)( 1e9 + 7) 


 

  
  double fx(double x ,double y){
    return x + y + 0.5 ;
  }
 void ahmed(){
    cout << fixed(10) ;
    double x0 , y0 , h , a , b ;
    cout << "fx(x0) = y0 , please enter x0 and y0 :" << nl ;
    cin >> x0 >> y0 ;
    cout << "please , enter interval : " << nl ; 
    cin >> a >> b ;
    cout << "please , enter step " << nl ;
    cin >> h ;
    vector < double > x , y ;
    for ( double k = x0 ; k <= b ; k += h )
    x.push_back(k) ;

    y.push_back(y0);
    int i ; 
    for ( i = 1 ; i < x.size() ; i++){
        y.push_back(y[i-1] + h * fx(x[i-1] , y[i-1])) ;
    }

    //------------------------------------------------------------
    // modified oylar 

    vector < double > ye1 , yi ;
    yi.push_back(y0) ;
    for ( i = 0 ; i < x.size() ; i++){
       if(i != x.size() - 1) {ye1.push_back(yi[i] + h * fx(x[i],yi[i])) ; }
       if(i) yi.push_back( yi[i-1] + (h / 2 ) * abs(fx(x[i-1],yi[i-1]) + fx(x[i] , ye1[i-1] ))) ;
    }
    i-- ;
    cout << nl ;
    cout << "by using modified  method : " << nl ;
    cout << "at x = " << x[i] << ", y = " << y[i] << nl ;
    cout << nl ;
    cout << "by using modified oylar method : " << nl ;
    cout << "at x = " << x[i] << " , y = " << yi[i] << nl ;
 }
 
 int main(){
  
  ll t = 1 ;
  // cin >> t ;
  while(t--){
    ahmed() ;
  }

  return 0 ;
}

